### PR TITLE
immediate rewards / terminal outcomes dual targets

### PR DESCRIPTION
**Pull request by YuriCat**
To distinguish both final results and immediate reward in each step,
both value and return estimation are computed by neural network and trainied separately.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -182,18 +203,20 @@ def compose_losses(policies, values, log_selected_policies, advantages, value_ta
 
     losses['p'] = (-log_selected_policies * turn_advantages).sum()
     losses['v'] = ((values - value_targets) ** 2).mul(vmasks).sum() / 2
+    if returns is not None:
+        losses['r'] = torch.abs(returns - return_targets).mul(vmasks).sum()
```
Comment (ikki407) [inline train.py:207]: IMHO: `smooth_l1_loss` (suggestion)

**Comment by ikki407** [comment]
nice update!

File: train.py (hunk 1 above)
Comment (YuriCat) [reply train.py:207]: good suggestion

Comment (YuriCat) [reply train.py]: applied

File: handyrl/model.py [hunk 2]
```diff
@@ -244,14 +244,14 @@ def inference(self, x, hidden, **kwargs):
             outputs = self.forward(xt, ht, **kwargs)
 
         return tuple(
-            [to_numpy(o).squeeze(0) for o in outputs[:-1]] +
+            [(to_numpy(o).squeeze(0) if o is not None else None) for o in outputs[:-1]] + \
```
Comment (ikki407) [inline handyrl/model.py:247]: `\` should be removed?

File: handyrl/train.py [hunk 3]
```diff
@@ -179,19 +201,22 @@ def compose_losses(policies, values, log_selected_policies, advantages, value_ta
...
+        losses['v'] = ((values - value_targets) ** 2).mul(vmasks).sum() / 2
+    if returns is not None:
+        losses['r'] = F.smooth_l1_loss(returns, return_targets, reduction='none').mul(vmasks).sum()
 
     entropy = dist.Categorical(logits=policies).entropy().mul(tmasks.sum(-1))
     losses['ent'] = entropy.sum()
 
-    loss_base = losses['p'] + losses['v']
+    loss_base = losses['p'] + losses.get('v', 0) + losses.get('r', 0)
     losses['total'] = loss_base + entropy.mul(1 - progress * 0.9).sum() * -entropy_regularization
```
Comment (ikki407) [inline handyrl/train.py:213]: The coefficient of `progress`, `0.9`, should be set in configuration?
What do you think?

File: handyrl/train.py [hunk 4]
```diff
@@ -204,59 +229,113 @@ def vtrace_base(batch, model, hidden, args):
...
+
+    if values_nograd is not None:
+        if values_nograd.size(2) == 2:  # two player zerosum game
+            values_nograd_opponent = -torch.stack([values_nograd[:, :, 1], values_nograd[:, :, 0]], dim=-1)
+            if args['observation']:
+                values_nograd = (values_nograd + values_nograd_opponent) / 2
+            else:
+                values_nograd = values_nograd + values_nograd_opponent
+                # Be careful, vmask in batch is changed here
+                batch['vmask'] = batch['vmask'].sum(-1, keepdim=True)
```
Comment (ikki407) [inline handyrl/train.py:243]: Is there anything that users should be careful?

File: handyrl/train.py (hunk 3 above)
Comment (YuriCat) [reply handyrl/train.py:213]: agree.
There might be better function than linear one.

Comment (ikki407) [reply handyrl/train.py:213]: It's difficult problem because it depends on games.
But I think the adjustment `1 - progress * 0.9` makes sense. So let's just keep this!

File: handyrl/model.py (hunk 2 above)
Comment (YuriCat) [reply handyrl/model.py]: yes. removed

**Comment by ikki407** [comment]
Reviewed
